### PR TITLE
Add wikibase bundle image with OAuth

### PR DESCRIPTION
**Pull request by tarrow**
Move the existing wikibase 1.30 image into a base folder for just vanilla
wikibase.

Create bundle image based off wikibase base and include OAuth in DockerFile
Also enable it in the LocalSettings.php template

Thoughts: 

- Should bundle be the default image in the example compose? I think yes because it makes having the 'full' setup running easier for people new to docker. People who are experienced enough to want vanilla wikibase can probably figure it out.

- Should we just append to LocalSettings.php.template rather than overwrite? Overwriting seems more explicit but perhaps just appending would be cleaner.

- Depending on how the automated build adds tags we should include new entries in wikibase/README.md

### CODE REVIEW

File: wikibase/1.30/bundle/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+RUN apt-get update && \
+    apt-get install --yes --no-install-recommends unzip
+
+ADD https://extdist.wmflabs.org/dist/extensions/OAuth-REL1_30-352260e.tar.gz .
+
+# Creates /mediawiki-extensions-Wikibase-REL1_30
+RUN tar xzf OAuth-REL1_30-352260e.tar.gz && rm *.tar.gz
+
+
+FROM wikibase/wikibase:1.30
```
Comment (addshore) [inline wikibase/1.30/bundle/Dockerfile]: I'm not sure how docker cloud and this build dependency will actually work out, might be worth doing some investigation.

AFAIK this bundle image and the base image will likely get built simultaneously and thus updates to the base image wouldn't be reflected in the bundle image on this build :/

File: wikibase/1.30/bundle/LocalSettings.php.template [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+# Wikibase
+require_once "${DOLLAR}IP/extensions/Wikibase/vendor/autoload.php";
+require_once "${DOLLAR}IP/extensions/Wikibase/lib/WikibaseLib.php";
+require_once "${DOLLAR}IP/extensions/Wikibase/repo/Wikibase.php";
+require_once "${DOLLAR}IP/extensions/Wikibase/repo/ExampleSettings.php";
+require_once "${DOLLAR}IP/extensions/Wikibase/client/WikibaseClient.php";
+require_once "${DOLLAR}IP/extensions/Wikibase/client/ExampleSettings.php";
+
+#OAuth
```
Comment (addshore) [inline wikibase/1.30/bundle/LocalSettings.php.template]: terribly nit pickey, but # OAuth :)

**Comment by addshore** [comment]
> **Should we just append to LocalSettings.php.template rather than overwrite? Overwriting seems more explicit but perhaps just appending would be cleaner.

Good point, appending could make sense.
Or some sort of magical DIR that can contain config snippets, all of which get loaded, but that might be a bit evil.

> Depending on how the automated build adds tags we should include new entries in wikibase/README.md

Yes README would need updating once we figure out what the tags will be etc :)

**Comment by addshore** [comment]
This seems to only be building ci/dockercloud (/wikibase/1.29)  now ;)

**Comment by addshore** [comment]
So, I just had another thought, it might be good to do this in 2 steps.

1) Create the name images & directories as exact copies of the current ones. So we have a -bundle image for each version, but there is nothing extra actually bundled in their yet
2) Then we make the modifications and add extra extensions once the files have moved around and building is already done.

It might also be worth opening up a dialogue with WMF people about moving the building to the WMF infra, then we could use the 'cool' docker-pkg script etc

File: wikibase/1.30/bundle/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,16 @@
+FROM ubuntu as fetcher
+
+RUN apt-get update && \
+    apt-get install --yes --no-install-recommends unzip
+
+ADD https://extdist.wmflabs.org/dist/extensions/OAuth-REL1_30-352260e.tar.gz .
```
Comment (legoktm) [inline wikibase/1.30/bundle/Dockerfile]: These URLs are not permanent and will change on you. Specifically, if someone backports a change to REL1_30, the sha1 will change, so the URL does too. You can use <https://www.mediawiki.org/w/api.php?action=query&list=extdistbranches&edbexts=OAuth&formatversion=2&format=json> to get the correct URL for REL1_30.

Comment (tarrow) [reply wikibase/1.30/bundle/Dockerfile]: Awesome, thanks; that's really helpful to know!

File: wikibase/1.30/bundle/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+RUN apt-get update && \
+    apt-get install --yes --no-install-recommends unzip jq curl ca-certificates
+ADD download-extension.sh .
+RUN bash download-extension.sh OAuth
+RUN tar xzf OAuth.tar.gz && rm *.tar.gz
+
+FROM wikibase/wikibase:1.30
+COPY --from=fetcher /OAuth /var/www/html/extensions/OAuth
+COPY LocalSettings.php.template /LocalSettings.php.template.additional
```
Comment (addshore) [inline wikibase/1.30/bundle/Dockerfile]: Perhaps this file should have a name specifying that it is from the bundle image?
LocalSettings.php.wikibase-bundle.template ? or something

File: wikibase/1.30/bundle/download-extension.sh [hunk 5]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/env bash
```
Comment (addshore) [inline wikibase/1.30/bundle/download-extension.sh]: Looking forward, this should either be download-oauth-extension. or download-extensions,? As we will either add more extensions to this file or add more files?

File: wikibase/1.30/bundle/Dockerfile (hunk 4 above)
Comment (tarrow) [reply wikibase/1.30/bundle/Dockerfile]: Sure

File: wikibase/1.30/bundle/download-extension.sh (hunk 5 above)
Comment (tarrow) [reply wikibase/1.30/bundle/download-extension.sh]: This is a generic script, it should be possible to call it again with a different parameter to a different single extension

Comment (addshore) [reply wikibase/1.30/bundle/download-extension.sh]: Ahh yes, so currently it still has OAuth hard coded:

| jq -r '.query.extdistbranches.extensions.OAuth.REL1_30

File: wikibase/README.md [hunk 6]
```diff
@@ -9,10 +9,16 @@ Automated build.
...
 ### Tags
 
-Image name                                  | Parent image            
-------------------------------------------- | ------------------------
-`wikibase/wikibase` : `latest`, `1.30`      | [mediawiki:1.30](https://hub.docker.com/_/mediawiki/)
-`wikibase/wikibase` : `legacy`, `1.29`      | [mediawiki:1.29](https://hub.docker.com/_/mediawiki/)
+Image name                                                       | Parent image
+---------------------------------------------------------------- | ------------------------
+`wikibase/wikibase` : `latest`, `1.30`                           | [mediawiki:1.30](https://hub.docker.com/_/mediawiki/)
+`wikibase/wikibase-base` : `latest`, `1.30` (identical to above) | [mediawiki:1.30](https://hub.docker.com/_/mediawiki/)
```
Comment (addshore) [inline wikibase/README.md]: This is actually wrong, -base is part of the tag name, not part of the image name. Same for -bundle.

File: wikibase/README.md [hunk 7]
```diff
@@ -9,10 +9,16 @@ Automated build.
 
 ### Tags
 
-Image name                                  | Parent image            
-------------------------------------------- | ------------------------
-`wikibase/wikibase` : `latest`, `1.30`      | [mediawiki:1.30](https://hub.docker.com/_/mediawiki/)
-`wikibase/wikibase` : `legacy`, `1.29`      | [mediawiki:1.29](https://hub.docker.com/_/mediawiki/)
+Image name                                                       | Parent image
+---------------------------------------------------------------- | ------------------------
+`wikibase/wikibase` : `latest`, `1.30`                           | [mediawiki:1.30](https://hub.docker.com/_/mediawiki/)
```
Comment (addshore) [inline wikibase/README.md]: This url is also wrong, this links to the mediawiki image, but should link to the wikibase/wikibase image..

Comment (tarrow) [reply wikibase/README.md]: I think this one is actually right; the *parent* image is mediawiki

File: wikibase/README.md (hunk 6 above)
Comment (tarrow) [reply wikibase/README.md]: Cool, fixing

**Pull request merged**